pkg/resources/masking_policy.go

```diff
@@ -2,282 +2,339 @@ package resources
-	ctx := context.Background()
-	objectIdentifier := sdk.NewSchemaObjectIdentifier(databaseName, schemaName, name)
-
-	signatureList := d.Get("signature").([]interface{})
-	signature := []sdk.TableColumnSignature{}
-	for _, s := range signatureList {
-		m := s.(map[string]interface{})
-		columns := m["column"].([]interface{})
-		for _, c := range columns {
-			cm := c.(map[string]interface{})
-			dt, err := sdk.ToDataType(cm["type"].(string))
-			if err != nil {
-				return err
-			}
-			signature = append(signature, sdk.TableColumnSignature{
-				Name: cm["name"].(string),
-				Type: dt,
-			})
+	arguments := d.Get("argument").([]any)
+	args := make([]sdk.TableColumnSignature, 0)
+	for _, arg := range arguments {
+		v := arg.(map[string]any)
+		dataType, err := sdk.ToDataType(v["type"].(string))
+		if err != nil {
+			return diag.FromErr(err)
 		}
+		args = append(args, sdk.TableColumnSignature{
+			Name: v["name"].(string),
+			Type: dataType,
+		})
 	}
 
 	returns, err := sdk.ToDataType(returnDataType)
 	if err != nil {
-		return err
+		return diag.FromErr(err)
 	}
+
+	// set optionals
 	opts := &sdk.CreateMaskingPolicyOptions{}
 	if comment, ok := d.Get("comment").(string); ok {
 		opts.Comment = sdk.String(comment)
 	}
-	if exemptOtherPolicies := d.Get("exempt_other_policies").(bool); exemptOtherPolicies {
-		opts.ExemptOtherPolicies = sdk.Bool(exemptOtherPolicies)
+	if v := d.Get("exempt_other_policies").(string); v != BooleanDefault {
+		parsed, err := strconv.ParseBool(v)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		opts.ExemptOtherPolicies = sdk.Pointer(parsed)
 	}
 
-	err = client.MaskingPolicies.Create(ctx, objectIdentifier, signature, returns, expression, opts)
+	err = client.MaskingPolicies.Create(ctx, id, args, returns, expression, opts)
 	if err != nil {
-		return err
+		return diag.FromErr(err)
 	}
-	d.SetId(helpers.EncodeSnowflakeID(objectIdentifier))
+	d.SetId(helpers.EncodeResourceIdentifier(id))
 
-	return ReadMaskingPolicy(d, meta)
+	return ReadMaskingPolicy(false)(ctx, d, meta)
 }
 
-// ReadMaskingPolicy implements schema.ReadFunc.
-func ReadMaskingPolicy(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*provider.Context).Client
-	id := helpers.DecodeSnowflakeID(d.Id()).(sdk.SchemaObjectIdentifier)
-
-	ctx := context.Background()
-	maskingPolicy, err := client.MaskingPolicies.ShowByID(ctx, id)
-	if err != nil {
-		return err
-	}
-	if err := d.Set(FullyQualifiedNameAttributeName, id.FullyQualifiedName()); err != nil {
-		return err
-	}
+func ReadMaskingPolicy(withExternalChangesMarking bool) schema.ReadContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+		client := meta.(*provider.Context).Client
+		id, err := sdk.ParseSchemaObjectIdentifier(d.Id())
+		if err != nil {
+			return diag.FromErr(err)
+		}
 
-	if err := d.Set("name", maskingPolicy.Name); err != nil {
-		return err
-	}
+		maskingPolicy, err := client.MaskingPolicies.ShowByID(ctx, id)
```

Comment:
Could you check ShowByID implementation if it's using collections.FindFirst? I think in main it's not using it and the `if errors.Is(err, sdk.ErrObjectNotFound)` branch will be never entered.

Reply:
Good point, I changed the sdk implementation to use collections.FindFirst.